File: typescript/src/filesIndex.ts
/**
 * L3 `files` index client — one metadata row per uploaded object.
 * Called after a successful R2 put(); best-effort.
 */

import { getConfig } from "./config";
import type { StoredFile } from "./envelope";
import { authHeaders, postJson } from "./transport";

export async function recordFile(
  stored: StoredFile,
  opts: {
    workspaceId: string;
    userId?: string;
    kind: string;
    entityRef?: { type?: string; id?: string };
  },
): Promise<string | undefined> {
  const cfg = getConfig();
  if (!cfg.filesEnabled || !cfg.platformApiUrl) return undefined;
  const resp = await postJson(
    `${cfg.platformApiUrl}/api/v1/files`,
    {
      workspace_id: opts.workspaceId,
      user_id: opts.userId,
      product: cfg.product,
      kind: opts.kind,
      key: stored.key,
      bucket: stored.bucket,
      public_url: stored.public_url,
      content_type: stored.content_type,
      size: stored.size,
      checksum: stored.checksum,
      entity_ref: opts.entityRef,
      created_at: stored.created_at,
    },
    { headers: authHeaders(cfg.platformToken), timeoutMs: cfg.timeoutMs, label: "files" },
  );
  if (!resp) return undefined;
  try {
    const body: any = await resp.json();
    const row = body?.data ?? body;
    const id = row?.file_id ?? row?._id;
    if (id) stored.file_id = String(id);
    return stored.file_id;
  } catch {
    return undefined;
  }
}

File: go/files_index.go
package streamoid

import "context"

// RecordFileOptions carries RecordFile's tenancy/classification fields.
type RecordFileOptions struct {
	WorkspaceID string
	UserID      string
	// Kind classifies the file for the audit trail (e.g. "file-upload",
	// "image-export") -- free-form, not part of EventVerbs.
	Kind string
	// EntityRef optionally links the file to a domain entity, e.g.
	// {"type": "asset-export", "id": key}.
	EntityRef map[string]any
}

type recordFileRequest struct {
	Key         string         `json:"key"`
	Bucket      string         `json:"bucket"`
	PublicURL   string         `json:"public_url"`
	Size        int64          `json:"size"`
	ContentType string         `json:"content_type,omitempty"`
	WorkspaceID string         `json:"workspace_id"`
	UserID      string         `json:"user_id,omitempty"`
	Kind        string         `json:"kind"`
	EntityRef   map[string]any `json:"entity_ref,omitempty"`
}

type recordFileResponse struct {
	Data struct {
		ID string `json:"_id"`
	} `json:"data"`
}

// RecordFile upserts stored into unified-backend's files index (POST
// /api/v1/files, upsert by storage key) and, on success, sets
// stored.FileID from the response. Best-effort: soft-skips if no platform
// API URL is configured or FilesEnabled is false; any transport/HTTP error
// is returned to the caller to log (mirroring the Python/TS SDKs' "caller
// decides whether to await or fire-and-forget" split -- RecordFile itself
// never panics or breaks the upload, it just reports the failure).
func (c *Client) RecordFile(ctx context.Context, stored *StoredFile, opts RecordFileOptions) error {
	if c == nil || !c.cfg.FilesEnabled || !c.cfg.HasPlatform() {
		return nil
	}
	req := recordFileRequest{
		Key:         stored.Key,
		Bucket:      stored.Bucket,
		PublicURL:   stored.PublicURL,
		Size:        stored.Size,
		ContentType: stored.ContentType,
		WorkspaceID: opts.WorkspaceID,
		UserID:      opts.UserID,
		Kind:        opts.Kind,
		EntityRef:   opts.EntityRef,
	}
	var resp recordFileResponse
	url := c.cfg.PlatformAPIURL + "/api/v1/files"
	if err := c.postJSON(ctx, url, req, &resp); err != nil {
		return err
	}
	if resp.Data.ID != "" {
		stored.FileID = resp.Data.ID
	}
	return nil
}
